### ISSUE TITLE
ruby type dispatcher failure

### ISSUE THREAD

**Issue by karlnapf** [swig/swig#1199]
We have a function of the form
```
void foo(A* a);
void foo(B b);
```

In the type dispatcher part of the swig-generated C code, when calling `foo(b)` with argument of type `B`, then, inside the dispatcher of `foo`, the function `SWIG_ConvertPtr` is called on `b`, which is not a pointer.
We get the following error message

```no implicit conversion of nil into String (TypeError)```

In our typemap for `B`, the precedence for the passed typemap object is already set to `SWIG_TYPECHECK_POINTER` (which is `=0` i.e. the highest), and inside the generated code, the correct dispatcher is just below one for the pointer to `A`. Therefore, we cannot move our desired type dispatcher above this failing call. 

All this only happens in ruby, and not in: Python, Java, Octave, C#

Tried with swig versions 3.0.8 and 3.0.12

**Comment by wsfulton** [comment]
There are a few fixes around NULL handling for Ruby in the github master version which will go into swig-4.0.0. Have a look at the forthcoming changes in https://github.com/swig/swig/blob/master/CHANGES.current and try it out. If that doesn't fix it, you'll need to provide a small cut down example to replicate.

**Comment by karlnapf** [comment]
I can confirm this still happens when using 7fbdd21b4999e5a064b77f99919f4f0c242899db

**Comment by goatshriek** [comment]
I've been trying to reproduce this based on the shogun typemaps but have so far been unsuccessful. Would you be able to provide a standalone example to replicate the issue?

**Comment by karlnapf** [comment]
In case this helps, here is a shogun branch to reproduce the error. I will try to cut out a minimal version when I find some time, but this is how I can reproduce the error locally.
https://github.com/karlnapf/shogun/tree/feature/ruby_typemap_bug

Configure and compile with (takes a while, you will need some dependencies such as ply, ctags, swig, etc, could use the [shogun sdk docker image](https://hub.docker.com/r/shogun/shogun-dev/)).
```
mkdir build && cd build
cmake .. -DBUILD_META_EXAMPLES=On -DCMAKE_BUILD_TYPE=DEBUG -DENABLE_TESTING=On -DTRAVIS_DISABLE_UNIT_TESTS=On -DINTERFACE_PYTHON=On -DINTERFACE_RUBY=On -DENABLE_PROTOBUF=Off -DLICENSE_GPL_SHOGUN=Off -DUSE_SVMLIGHT=Off -DTRAVIS_DISABLE_META_CPP=On
make -j 4
```

Reproduce the error with
```
ctest -R generated_ruby-base_api-put_get_add -V
```
which should fail in the line that uses the heavily overloaded `put` method, `labels.put("labels", vector)`.

See that it works in Python via running
```
ctest -R generated_python-base_api-put_get_add -V
```

The method itself is implemented here:
https://github.com/karlnapf/shogun/blob/feature/ruby_typemap_bug/src/shogun/base/SGObject.h#L482
The corresponding swig template is instantiated here:
https://github.com/karlnapf/shogun/blob/feature/ruby_typemap_bug/src/interfaces/swig/shogun.i#L222
Our ruby typemap for `SGVector` is here: https://github.com/karlnapf/shogun/blob/feature/ruby_typemap_bug/src/interfaces/ruby/swig_typemaps.i#L17

The swig generated interface source file is at `build/src/interfaces/ruby/shogunRUBY_wrap.cxx`, in case 
 you want to gdb the dispatcher code (see the `ctest` details above to set up the environment).

**Comment by goatshriek** [comment]
Excellent, thank you very much for the detailed information! I will dig into this and see what I can turn up.

**Comment by vigsterkr** [comment]
@goatshriek @wsfulton i've added an easy fix for this bug see #1717 
basically this just guards the case when a nil is being passed to `StringValuePtr`
